[Merged by Bors] - feat(GroupTheory/GroupAction): new `FixedPoints` module with a few basic properties of `fixedBy`

adri326 opened the pull request:
Introduces a new module, `Mathlib.GroupTheory.GroupAction.FixedPoints`,
which contains some properties of `MulAction.fixedBy` that wouldn't quite belong to `Mathlib.GroupTheory.GroupAction.Basic`.

---

This is my first contribution ever to Mathlib, and the first contribution in a series of contributions to eventually merge [my formalization of Rubin's theorem](https://github.com/adri326/rubin-lean4/) into Lean.

I have a few questions around the design decisions that I took:

- ~~should `movedBy` be defined in `Mathlib.GroupTheory.GroupAction.Basic` instead?~~
- `Mathlib.GroupTheory.GroupAction.FixingSubgroup` uses `M` for a group, would it be out-of-scope for this PR to also replace it with `G`?
- should I try to further minimize the usage of tactic groups?
- I'm trying to translate the properties that I will need later for the formalization of Rubin's theorem, so I wonder if `not_commute_of_disjoint_movedBy_preimage` might be too out-of-scope

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

jcommelin commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,289 @@
+theorem fixedBy_subset_fixedBy_pow (g : G) (j : ℤ) :
+    fixedBy α g ⊆ fixedBy α (g^j) := by
+  intro a a_in_fixedBy
+  rw [
+    mem_fixedBy,
+    zpow_smul_eq_iff_minimalPeriod_dvd,
+    minimalPeriod_eq_one_of_fixedBy a_in_fixedBy
+  ]
```
```suggestion
  rw [mem_fixedBy, zpow_smul_eq_iff_minimalPeriod_dvd,
    minimalPeriod_eq_one_of_fixedBy a_in_fixedBy]
```

jcommelin commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,289 @@
+
+@[to_additive]
+theorem fixedBy_eq_compl_movedBy {m : M} : fixedBy α m = (movedBy α m)ᶜ := by
+  rw [<-compl_compl (fixedBy α m), movedBy_eq_compl_fixedBy]
```
```suggestion
  rw [← compl_compl (fixedBy α m), movedBy_eq_compl_fixedBy]
```

adri326 replied:
I noticed this style issue in the meantime, yeah

jcommelin commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,289 @@
+@[to_additive]
+theorem smul_mem_movedBy_iff_mem_movedBy {a : α} {g : G} :
+    g • a ∈ movedBy α g ↔ a ∈ movedBy α g := by
+  repeat rw [<-not_mem_fixedBy_iff_mem_movedBy]
```
```suggestion
  repeat rw [← not_mem_fixedBy_iff_mem_movedBy]
```

jcommelin commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,289 @@
+@[to_additive]
+theorem movedBy_mul (m₁ m₂ : M) : movedBy α (m₁ * m₂) ⊆ movedBy α m₁ ∪ movedBy α m₂ := by
+  repeat rw [movedBy_eq_compl_fixedBy]
+  rw [<-Set.compl_inter, Set.compl_subset_compl]
```
```suggestion
  rw [← Set.compl_inter, Set.compl_subset_compl]
```

jcommelin commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,289 @@
+@[to_additive]
+theorem smul_pow_subset_of_movedBy_subset {s t : Set α} {g : G}  (t_superset : movedBy α g ⊆ t)
+    (s_ss_t : s ⊆ t) (j : ℤ): (fun a => g^j • a) ⁻¹' s ⊆ t := by
+  rw [<-smul_pow_preimage_eq_of_movedBy_subset t_superset j]
```
```suggestion
  rw [← smul_pow_preimage_eq_of_movedBy_subset t_superset j]
```

jcommelin commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,289 @@
+    intro a
+    rw [one_smul]
+    by_contra ma_ne_a
+    rwa [<-ne_eq, <-mem_movedBy, moved_empty] at ma_ne_a
```
```suggestion
    rwa [← ne_eq, ← mem_movedBy, moved_empty] at ma_ne_a
```

jcommelin commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,289 @@
+    rwa [<-ne_eq, <-mem_movedBy, moved_empty] at ma_ne_a
+  ),
+  fun eq_one => eq_one.symm ▸ movedBy_one_eq_empty α M
+⟩
```
This closing bracket should move to the previous line.
Alternatively, consider starting the tactic proof (`by`-block) on line 263, and splitting into two parts with `constructor`.

adri326 replied:
I did the latter, since most of the proof was in a `by`-block anyway

adri326 commented:
I fixed the above issues; the CI should pass, as I haven't changed any public-facing interfaces.

riccardobrasca commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,286 @@
+
+/-- In a multiplicative group action, the points fixed by `g` are also fixed by `g⁻¹` -/
+@[to_additive "In an additive group action, the points fixed by `g` are also fixed by `g⁻¹`"]
+theorem fixedBy_eq_fixedBy_inv (g : G) : fixedBy α g = fixedBy α g⁻¹ := by
```
Should `α` be explicit here?

adri326 replied:
It should, yes, although someone should probably do an automated pass on the entirety of mathlib because I've had to manually force-provide arguments several times

riccardobrasca commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,286 @@
+  rfl
+
+@[to_additive]
+theorem minimalPeriod_eq_one_of_fixedBy {a : α} {g : G} (a_in_fixedBy : a ∈ fixedBy α g) :
```
Can you please check that the name generated by `to_additive` is correct?

adri326 replied:
I'll do that, although I think the name should be the same, it will just be in the `AddAction` context.

riccardobrasca replied:
I just wanted to check that the `one` doesn't become `zero`.

adri326 replied:
Oh, good point!

adri326 replied:
You were right, it substituted it with a `zero`

riccardobrasca commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,286 @@
+  exact a_in_fixedBy
+
+@[to_additive]
+theorem fixedBy_subset_fixedBy_pow (g : G) (j : ℤ) :
```
Same here, should `α` be explicit?

riccardobrasca commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,286 @@
+variable (M)
+
+@[to_additive (attr := simp)]
+theorem fixedBy_one_eq_univ :
```
Same.

riccardobrasca commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,286 @@
+  rw [Nat.cast_one]
+  exact one_dvd j
+
+variable (M)
```
You can write
```lean
variable (M) in
```
just before a lemma to avoid to make `M` explicit and then implicit again.

riccardobrasca commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,286 @@
+
+@[to_additive]
+theorem fixedBy_subset_fixedBy_pow (g : G) (j : ℤ) :
+    fixedBy α g ⊆ fixedBy α (g^j) := by
```
```suggestion
    fixedBy α g ⊆ fixedBy α (g ^ j) := by
```
And similar elsewhere (not very important of course, but I think this style is more common in mathlib).

riccardobrasca commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,286 @@
+@[to_additive "If `s` is a superset of `movedBy α g`, then `g` cannot move a point outside of `s`."]
+theorem smul_pow_mem_of_movedBy_subset {a : α} {s : Set α} {g : G} (s_subset : movedBy α g ⊆ s)
+    (j : ℤ) : g^j • a ∈ s ↔ a ∈ s := by
+  rcases (Classical.em (a ∈ movedBy α (g^j))) with a_moved | a_fixed
```
The tactic `by_cases` does exactly this and it is nicer.

adri326 replied:
I do miss the convenience of NNG's `cases`, since I never remember the name of the generated subgoals, and you need to name them explicitely to be able to access the generated hypothesis

riccardobrasca replied:
I mean you do this
```lean
/-- If `movedBy α g ⊆ s`, then `g` cannot move a point of `s` outside of `s`. -/
@[to_additive "If `movedBy α g ⊆ s`, then `g` cannot move a point of `s` outside of `s`."]
theorem smul_zpow_mem_of_movedBy_subset {a : α} {s : Set α} {g : G} (s_subset : movedBy α g ⊆ s)
    (j : ℤ) : g ^ j • a ∈ s ↔ a ∈ s := by
  by_cases h :  a ∈ movedBy α (g ^ j)
  · constructor
    all_goals (intro; apply s_subset)
    all_goals apply movedBy_zpow_subset_movedBy g j
    · exact h
    · rw [smul_mem_movedBy_iff_mem_movedBy]
      exact h
  · rw [not_mem_movedBy_iff_mem_fixedBy, mem_fixedBy] at h
    rw [h]
```

riccardobrasca commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,286 @@
+  ext a
+  rw [Set.mem_preimage, smul_pow_mem_of_movedBy_subset superset]
+
+/-- If `movedBy α g ⊆ t`, then `g` cannot send a subset of `t` outside of `t` -/
```
```suggestion
/-- If `movedBy α g ⊆ t`, then `g` cannot send a subset of `t` outside of `t`. -/
```

riccardobrasca commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,286 @@
+  rw [Set.mem_preimage, smul_pow_mem_of_movedBy_subset superset]
+
+/-- If `movedBy α g ⊆ t`, then `g` cannot send a subset of `t` outside of `t` -/
+@[to_additive]
```
I don't remember if we have to specify the docstring to `to_additive`, can you please check this?

Same below. If not then above you repeated the docstring.

adri326 replied:
We do, yeah

adri326 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,331 @@
+variable {G : Type u} [AddGroup G] [AddAction G α]
+
+theorem _root_.AddAction.minimalPeriod_eq_one_of_fixedBy {a : α} {g : G}
+  (a_in_fixedBy : a ∈ AddAction.fixedBy α g) :
```
I'm tired

adri326 commented:
I fixed the issues pointed out by @riccardobrasca, as well as some I noticed in the meantime (some theorems were named `pow` when they referred to `zpow`).

I would like to say that trying to get this first pull request merged has been a pretty demotivating experience so far:
there is no proper linter or formatter that I could run locally,
building documentation locally is extremely slow and the instructions for doing so are poorly documented, etc.

The backlog of pull requests means that each round of small issue that could be caught ahead of time with proper tooling instead adds at least a day before this pull request gets approved. I understand the need to have PRs be properly reviewed, but I long for tools to reduce these reviews down to the more impactful issues.

riccardobrasca commented:
> I fixed the issues pointed out by @riccardobrasca, as well as some I noticed in the meantime (some theorems were named `pow` when they referred to `zpow`).
> 
> I would like to say that trying to get this first pull request merged has been a pretty demotivating experience so far: there is no proper linter or formatter that I could run locally, building documentation locally is extremely slow and the instructions for doing so are poorly documented, etc.
> 
> The backlog of pull requests means that each round of small issue that could be caught ahead of time with proper tooling instead adds at least a day before this pull request gets approved. I understand the need to have PRs be properly reviewed, but I long for tools to reduce these reviews down to the more impactful issues.

I am sorry this has been a painful experience.

One thing to keep in mind is that last week there was a pretty important Lean conference, so a lot of people didn't have time to review PRs: we know that it is a slow process, but maybe you were just unlucky.

Can you please copy/paste this message on Zulip? It is very possible that some of tools you want exist, but are maybe not very well documented (for example you can run all the linters locally, it is just slow).

urkud commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,331 @@
+variable (α) in
+/-- In a multiplicative group action, the points fixed by `g` are also fixed by `g⁻¹` -/
+@[to_additive "In an additive group action, the points fixed by `g` are also fixed by `g⁻¹`"]
+theorem fixedBy_eq_fixedBy_inv (g : G) : fixedBy α g = fixedBy α g⁻¹ := by
```
Please swap LHS with RHS and mark it as a `simp` lemma.

adri326 replied:
Done, I avoided introducing new simp lemmas but this one is straightforward enough

urkud commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,331 @@
+  rw [Set.preimage_compl, fixedBy_conj]
+
+@[to_additive]
+theorem movedBy_conj' {g h : G} :
```
Naming convention says that this should be called `smul_movedBy`.

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,331 @@
+open Pointwise
+
+universe u v
+variable {α : Type v}
+variable {G : Type u} [Group G] [MulAction G α]
+variable {M : Type u} [Monoid M] [MulAction M α]
```
Someone can correct me if I'm wrong, but I think `Type*` is preferred over explicit universes?

adri326 replied:
I can't find where this is said, but from a quick ripgrep there seems to be a lot more instances of `Type*` than `universe`

urkud replied:
With these variables, if you ever use `M` and `G` in the same theorem, then it won't be universe polymorphic.

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,331 @@
+@[to_additive "In an additive group action, the points fixed by `g` are also fixed by `g⁻¹`"]
+theorem fixedBy_eq_fixedBy_inv (g : G) : fixedBy α g = fixedBy α g⁻¹ := by
+  ext x
+  repeat rw [mem_fixedBy]
+  constructor
+  all_goals (intro gx_eq_x; nth_rw 1 [← gx_eq_x])
+  · exact inv_smul_smul g x
+  · rw [← mul_smul, mul_right_inv, one_smul]
```
```suggestion
  rw [mem_fixedBy, mem_fixedBy, inv_smul_eq_iff, eq_comm]
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,331 @@
+
+theorem minimalPeriod_eq_one_of_fixedBy {a : α} {g : G} (a_in_fixedBy : a ∈ fixedBy α g) :
+    Function.minimalPeriod (fun x => g • x) a = 1 := by
+  rw [Function.minimalPeriod_eq_one_iff_isFixedPt]
+  exact a_in_fixedBy
```
```suggestion
  rwa [Function.minimalPeriod_eq_one_iff_isFixedPt]
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,331 @@
+theorem _root_.AddAction.minimalPeriod_eq_one_of_fixedBy {a : α} {g : G}
+    (a_in_fixedBy : a ∈ AddAction.fixedBy α g) :
+    Function.minimalPeriod (fun x => g +ᵥ x) a = 1 := by
+  rw [Function.minimalPeriod_eq_one_iff_isFixedPt]
+  exact a_in_fixedBy
```
```suggestion
  rwa [Function.minimalPeriod_eq_one_iff_isFixedPt]
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,331 @@
+theorem fixedBy_subset_fixedBy_zpow (g : G) (j : ℤ) :
+    fixedBy α g ⊆ fixedBy α (g ^ j) := by
+  intro a a_in_fixedBy
+  rw [mem_fixedBy, zpow_smul_eq_iff_minimalPeriod_dvd,
+    minimalPeriod_eq_one_of_fixedBy a_in_fixedBy]
+  rw [Nat.cast_one]
```
```suggestion
  rw [mem_fixedBy, zpow_smul_eq_iff_minimalPeriod_dvd,
    minimalPeriod_eq_one_of_fixedBy a_in_fixedBy, Nat.cast_one]
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,331 @@
+theorem fixedBy_one_eq_univ :
+    fixedBy α (1 : M) = Set.univ := by
+  ext a
+  rw [mem_fixedBy, one_smul]
+  exact ⟨fun _ => Set.mem_univ a, fun _ => rfl⟩
```
```suggestion
  simp
```
or
```suggestions
  simp_rw [mem_fixedBy, one_smul, Set.mem_univ]
```

adri326 replied:
This one I'm unsure if I agree; I can just have `Set.eq_univ_iff_forall.mpr (fun a => one_smul M a)`, though

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,331 @@
+@[to_additive "If `movedBy α g ⊆ s`, then `g` cannot move a point of `s` outside of `s`."]
+theorem smul_zpow_mem_of_movedBy_subset {a : α} {s : Set α} {g : G} (s_subset : movedBy α g ⊆ s)
+    (j : ℤ) : g ^ j • a ∈ s ↔ a ∈ s := by
+  by_cases a ∈ movedBy α (g ^ j)
+  case pos a_moved =>
+    constructor
+    all_goals (intro; apply s_subset)
+    all_goals apply movedBy_zpow_subset_movedBy g j
+    · exact a_moved
+    · rw [smul_mem_movedBy_iff_mem_movedBy]
+      exact a_moved
+  case neg a_fixed =>
+    rw [not_mem_movedBy_iff_mem_fixedBy, mem_fixedBy] at a_fixed
+    rw [a_fixed]
```
Here's an alternate proof, not because I'm suggesting you adopt it, but because it indicates to me that the "substance" of this lemma is `g • s = s` (`g` takes `s` to `s`). So I wonder whether some of your later results might be better phrased in this pointwise manner. For instance, `smul_zpow_mem_fixedBy_of_commute` could be phrased as `Commute g h → h • fixedBy α g = fixedBy α g`.

```
have h1 : g • s = s
  · ext a
    by_cases h : g • a = a
    · rw [Set.mem_smul_set_iff_inv_smul_mem, ← h, inv_smul_smul, h]
    rw [Set.mem_smul_set_iff_inv_smul_mem]
    exact iff_of_true (s_subset (smul_inv_mem_movedBy_iff_mem_movedBy.mpr h)) (s_subset h)
  have h2 : g ^ j • s = s
  · exact fixedBy_subset_fixedBy_zpow (Set α) g j h1
  rw [← Set.ext_iff.mp h2 (g ^ j • a), Set.smul_mem_smul_set_iff]
```

adri326 replied:
You are right, if I go with the pointwise smul route, then I can re-use a lot of the theorems from above (since `g • s = s` is definitionally equal to `s ∈ fixedBy (Set α) g`). I was wary of using pointwise smul, because they are used rarely enough that I didn't know about it for a while at first.

semorrison commented:
> (for example you can run all the linters locally, it is just slow).

If you have enough local compute `make lint` is all that it takes to run the linter locally.

Automatic code formatting is definitely on the roadmap!

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,322 @@
+variable (α) in
+/-- The set of points moved by an element of the action. -/
+@[to_additive "The set of points moved by an element of the action."]
+def movedBy (m : M) : Set α := { a : α | m • a ≠ a }
```
Is just writing `(fixedBy α m)ᶜ` everywhere an option, or are some things too painful?

adri326 replied:
There are a few works out there that chain operations with this structure, which they usually call the "support" of the action; this is the case in Rubin's theorem. I don't mind changing this to an `abbrev` or `notation`, though, and that would probably allow me to drop a few theorems

adri326 commented:
I replaced `movedBy α g` with `(fixedBy α g)ᶜ`, and added some helper lemmas around `s ∈ fixedBy (Set α) g`. I tried having `abbrev` but it made `rw` only work sporadically, and I couldn't get `notation` to work like I wanted to.

adri326 commented:
@tb65536 haven't heard back in a little while, are there other design patterns that are problematic with this PR?

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,292 @@
+    g⁻¹ • a ∈ fixedBy α g ↔ a ∈ fixedBy α g := by
+  rw [← fixedBy_inv_eq_fixedBy, smul_mem_fixedBy_iff_mem_fixedBy, fixedBy_inv_eq_fixedBy]
+
+@[to_additive minimalPeriod_eq_one_of_fixedBy]
+theorem minimalPeriod_eq_one_of_fixedBy {a : α} {g : G} (a_in_fixedBy : a ∈ fixedBy α g) :
+    Function.minimalPeriod (fun x => g • x) a = 1 := by
+  rwa [Function.minimalPeriod_eq_one_iff_isFixedPt]
```
Maybe upgrade this to an `iff`?
```suggestion
@[to_additive minimalPeriod_eq_one_iff_fixedBy]
theorem minimalPeriod_eq_one_iff_fixedBy {a : α} {g : G} :
    Function.minimalPeriod (fun x => g • x) a = 1 ↔ a ∈ fixedBy α g :=
  Function.minimalPeriod_eq_one_iff_isFixedPt
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,292 @@
+variable (M α) in
+@[to_additive (attr := simp)]
+theorem fixedBy_one_eq_univ : fixedBy α (1 : M) = Set.univ :=
+  Set.eq_univ_iff_forall.mpr (fun a => one_smul M a)
```
Minor golf:
```suggestion
  Set.eq_univ_iff_forall.mpr (one_smul M)
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,292 @@
+
+variable (α) in
+@[to_additive]
+theorem fixedBy_conj (g h : G) :
+    fixedBy α (h * g * h⁻¹) = (fun a => h⁻¹ • a) ⁻¹' fixedBy α g := by
+  ext a
+  rw [Set.mem_preimage, mem_fixedBy, mem_fixedBy]
```
I would suggest just cleaning up the statement of this theorem and deleting one or the other.
```suggestion
theorem fixedBy_conj (g h : G) : fixedBy α (h * g * h⁻¹) = h • fixedBy α g := by
  ext a
  rw [Set.mem_smul_set_iff_inv_smul_mem, mem_fixedBy, mem_fixedBy]
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,292 @@
+`s` after being moved by `g`."]
+theorem set_mem_fixedBy_iff (s : Set α) (g : G) :
+    s ∈ fixedBy (Set α) g ↔ ∀ x, x ∈ s ↔ g • x ∈ s := by
+  constructor
+  · intro fixed x
+    refine ⟨
+      fun x_in_s => fixed ▸ Set.smul_mem_smul_set x_in_s,
+      fun gx_in_s => ?inv_in_set⟩
+    rwa [← fixed, Set.smul_mem_smul_set_iff] at gx_in_s
+  · intro closed
+    rw [← fixedBy_inv_eq_fixedBy]
+    ext x
+    rw [Set.mem_inv_smul_set_iff]
+    exact ⟨fun gxs => (closed x).mpr gxs, fun xs => (closed x).mp xs⟩
```
```suggestion
  simp_rw [mem_fixedBy, ← eq_inv_smul_iff, Set.ext_iff, Set.mem_inv_smul_set_iff]
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,292 @@
+    exact ⟨fun gxs => (closed x).mpr gxs, fun xs => (closed x).mp xs⟩
+
+theorem smul_mem_of_set_mem_fixedBy {s : Set α} {g : G} (s_in_fixedBy : s ∈ fixedBy (Set α) g)
+    {x : α} : x ∈ s ↔ g • x ∈ s := (set_mem_fixedBy_iff s g).mp s_in_fixedBy x
```
Usually we would put the more complicated expression on the left side of the `iff`, so that it `rw`'s naturally.

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,266 @@
+If `g` and `h` commute, then `g` fixes `(h^j) • x` iff `g` fixes `x`.
+-/
+@[to_additive "If `g` and `h` commute, then `g` fixes `(j • h) +ᵥ x` iff `g` fixes `x`."]
+theorem smul_zpow_fixedBy_eq_of_commute {g h : G} (comm : Commute g h) (j : ℤ):
```
```suggestion
theorem smul_zpow_fixedBy_eq_of_commute {g h : G} (comm : Commute g h) (j : ℤ) :
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,266 @@
+If `g` and `h` commute, then `g` moves `h^j • x` iff `g` moves `x`.
+-/
+@[to_additive "If `g` and `h` commute, then `g` moves `(j • h) +ᵥ x` iff `g` moves `x`."]
+theorem smul_zpow_movedBy_eq_of_commute {g h : G} (comm : Commute g h) (j : ℤ):
```
```suggestion
theorem smul_zpow_movedBy_eq_of_commute {g h : G} (comm : Commute g h) (j : ℤ) :
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,266 @@
+@[to_additive "If the additive action of `M` on `α` is faithful,
+then `fixedBy α m = Set.univ` implies that `m = 1`."]
+theorem fixedBy_eq_univ_iff_eq_one {m : M} : fixedBy α m = Set.univ ↔ m = 1 := by
+  refine ⟨fun moved_empty => ?eq_one, fun eq_one => eq_one ▸ fixedBy_one_eq_univ α M⟩
+  apply FaithfulSMul.eq_of_smul_eq_smul (α := α)
+  intro a
+  rw [one_smul]
+  by_contra ma_ne_a
+  rw [← mem_fixedBy, moved_empty] at ma_ne_a
+  exact ma_ne_a (Set.mem_univ a)
```
```suggestion
  rw [← (smul_left_injective' (M := M) (α := α)).eq_iff, Set.eq_univ_iff_forall]
  simp_rw [Function.funext_iff, one_smul, mem_fixedBy]
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,266 @@
+is disjoint from `(fixedBy α g)ᶜ`, then `g` and `h` cannot commute."]
+theorem not_commute_of_disjoint_movedBy_preimage {g h : G} (ne_one : g ≠ 1)
+    (disjoint : Disjoint (fixedBy α g)ᶜ (h • (fixedBy α g)ᶜ)) : ¬Commute g h := by
+  intro comm
+  rw [movedBy_mem_fixedBy_of_commute comm, disjoint_self, Set.bot_eq_empty, ← Set.compl_univ,
+    compl_inj_iff, fixedBy_eq_univ_iff_eq_one] at disjoint
+  exact ne_one disjoint
```
```suggestion
  contrapose! ne_one with comm
  rwa [movedBy_mem_fixedBy_of_commute comm, disjoint_self, Set.bot_eq_empty, Set.compl_empty_iff,
    fixedBy_eq_univ_iff_eq_one] at disjoint
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixingSubgroup.lean:
```diff
@@ -36,6 +37,10 @@ TODO :
 
+* add `to_additive` for the various lemmas
+
+* `movingSubgroup`
```
What is `movingSubgroup`?

tb65536 replied:
Ah, I see. It's the fixing subgroup of the complement. Are you still planning on adding this, or should this TODO be removed?

adri326 replied:
No, I'm not going to add it for the same reason that `movedBy` was scrapped

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixingSubgroup.lean:
```diff
@@ -120,6 +125,16 @@ theorem mem_fixingSubgroup_iff {s : Set α} {m : M} : m ∈ fixingSubgroup M s 
 
+theorem mem_fixingSubgroup_iff_subset_fixedBy {s : Set α} {m : M} :
+    m ∈ fixingSubgroup M s ↔ s ⊆ fixedBy α m := by
+  rw [mem_fixingSubgroup_iff, Set.subset_def]
+  simp only [mem_fixedBy]
```
```suggestion
  simp_rw [mem_fixingSubgroup_iff, Set.subset_def, mem_fixedBy]
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixingSubgroup.lean:
```diff
@@ -120,6 +125,16 @@ theorem mem_fixingSubgroup_iff {s : Set α} {m : M} : m ∈ fixingSubgroup M s 
+
+theorem mem_fixingSubgroup_compl_iff_movedBy_subset {s : Set α} {m : M} :
+    m ∈ fixingSubgroup M sᶜ ↔ (fixedBy α m)ᶜ ⊆ s := by
+  rw [mem_fixingSubgroup_iff_subset_fixedBy]
+  exact Set.compl_subset_comm
```
```suggestion
  rw [mem_fixingSubgroup_iff_subset_fixedBy, Set.compl_subset_comm]
```

tb65536 commented on Mathlib/GroupTheory/GroupAction/FixingSubgroup.lean:
```diff
@@ -161,4 +176,13 @@ theorem fixedPoints_subgroup_iSup {ι : Sort*} {P : ι → Subgroup M} :
 #align fixed_points_subgroup_supr fixedPoints_subgroup_iSup
 
+/-- The orbit of the fixing subgroup of `sᶜ` (ie. the moving subgroup of `s`) is a subset of `s` -/
+theorem orbit_fixingSubgroup_compl_subset {s : Set α}
+    {a : α} (a_in_s : a ∈ s) : MulAction.orbit (fixingSubgroup M sᶜ) a ⊆ s := by
```
```suggestion
theorem orbit_fixingSubgroup_compl_subset {s : Set α} {a : α} (a_in_s : a ∈ s) :
  MulAction.orbit (fixingSubgroup M sᶜ) a ⊆ s := by
```

tb65536 reviewed (approved):
Last really minor comments, but this is looking good to me

adri326 commented on Mathlib/GroupTheory/GroupAction/FixedPoints.lean:
```diff
@@ -0,0 +1,266 @@
+theorem movedBy_mem_fixedBy_of_commute {g h : G} (comm: Commute g h) :
+    (fixedBy α g)ᶜ ∈ fixedBy (Set α) h := by
+  rw [mem_fixedBy, Set.compl_eq_univ_diff, Set.smul_set_sdiff,
+    Set.smul_set_univ, fixedBy_mem_fixedBy_of_commute comm]
```
I ended up writing this lemma in the meantime, I probably have it in other PRs already. Might as well get that lemma merged now :)

tb65536 commented:
LGTM! Thanks for adding all this API

maintainer merge

tb65536 commented:
(I guess it should be delegated, since CI is still running)

adri326 commented:
Looks like the branch didn't move to the merge queue @tb65536

Vierkantor reviewed (approved):
Thanks!

bors merge
